File: ruby/0763-partition-labels.rb

def partition_labels(s)
    last_index = Hash.new(0)

    s.each_char.with_index do |char,index|
        last_index[char] = index
    end

    result = []
    size = s_end = 0

    s.each_char.with_index do |char,index|
        size +=1
        s_end = [last_index[char],s_end].max

        if (index == s_end)
            result.append(size)
            size =0
        end
    end

    return result

end

File: kotlin/0763-partition-labels.kt
class Solution {
    fun partitionLabels(s: String): List<Int> {
        if (s.length == 1) return listOf(1)
        val lastIndexOfChar = mutableMapOf<Char, Int>().apply {
            s.forEachIndexed { index, char -> this[char] = index }
        }
        val partitions = mutableListOf<Int>()
        var l = 0
        while (l in s.indices) {
            // keep moving l to find the first char that has a last occurrence greater than l
            while (l == lastIndexOfChar[s[l]]) {
                // if l is at the one-and-only occurrence of the char s[l], make a partition and move l
                partitions.add(1)
                l++
                if (l > s.lastIndex) return partitions
            }

            var r = lastIndexOfChar.getValue(s[l])
            var j = l + 1
            while (j <= r) {
                if (lastIndexOfChar.getValue(s[j]) <= r){
                    j++
                    continue
                }
                r = lastIndexOfChar.getValue(s[j])
                j++
            }
            partitions.add((r - l) + 1)
            l = r + 1
        }
        return partitions
    }
}